flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableDescriptor.java

```diff
@@ -0,0 +1,225 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.api;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.configuration.ConfigOption;
+import org.apache.flink.configuration.ConfigurationUtils;
+import org.apache.flink.table.catalog.CatalogTable;
+import org.apache.flink.table.factories.FactoryUtil;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * Describes a {@link CatalogTable} representing a source or sink.
+ *
+ * <p>This can be used to register a table in the Table API, see (TODO FLINK-23065).
+ *
+ * <p>This class closely resembles the corresponding SQL DDL structure and {@link CatalogTable}.
+ */
+@PublicEvolving
+public final class TableDescriptor {
+
+    private final Schema schema;
+    private final Map<String, String> options;
+    private final List<String> partitionKeys;
+    private final @Nullable String comment;
+
+    private TableDescriptor(
+            Schema schema,
+            Map<String, String> options,
+            List<String> partitionKeys,
+            @Nullable String comment) {
+        this.schema = schema;
+        this.options = Collections.unmodifiableMap(options);
+        this.partitionKeys = Collections.unmodifiableList(partitionKeys);
+        this.comment = comment;
+    }
+
+    /**
+     * Creates a new {@link Builder} for a table using the given connector.
+     *
+     * @param connector The factory identifier for the connector.
+     */
+    public static Builder forConnector(String connector) {
+        final Builder descriptorBuilder = new Builder();
+        descriptorBuilder.option(FactoryUtil.CONNECTOR, connector);
+
+        return descriptorBuilder;
+    }
+
+    public Schema getSchema() {
+        return schema;
+    }
+
+    public Map<String, String> getOptions() {
+        return options;
+    }
+
+    public List<String> getPartitionKeys() {
+        return partitionKeys;
+    }
+
+    public Optional<String> getComment() {
+        return Optional.ofNullable(comment);
+    }
+
+    // ---------------------------------------------------------------------------------------------
+
+    @Override
+    public boolean equals(Object obj) {
```

Comment:
implement a `toString` that shows a `CREATE TABLE` statement, similar how `ResolvedSchema` shows the schema part in SQL

Reply:
I agree that a proper `toString` is useful. Generating proper DDL isn't trivial, and some utilities for escaping aren't available here, either. I generate some pseudo-DDL for now and enclosed it in `[]` which I think is what we do for types somewhere, too.

Reply:
It doesn't need to be proper DDL, pseudo-DDL is fine. For `Schema` we also just use an approximate representation. But I think it reads nicely when working in a notebook or debugging session.